### PR TITLE
Overhaul - Context

### PR DESCRIPTION
**Pull request by jansule**
This example uses react's `context` feature to compose custom editors without the need of rewriting everything again.

The surrounding app would have to wrap a `<Composition.Context value=...>` component around the outermost geostyler component.
value would be an object e.g. like this:

```jsx
 composition: {
        LineEditor: {
          widthField: <CustomWidth />, //replace field with custom component
          colorField: false // deactivate field
        },
        Editor: {
          lineEditor: <CustomLineEditor />, // replace editor with custom component
          markEditor: false, // deactivate editor
        }
      }
```

**Update:** Set up customisation of editors/editor fields following this pattern. Now it is possible to replace or deactivate single fields within a symbolizerEditor. Also, complete editors are replacable. Take a look at `CompositionContext.tsx` to get an idea of all customisable components.

@KaiVolland @chrismayer please review

### CODE REVIEW

File: src/Component/Symbolizer/LineEditor/LineEditor.tsx [hunk 1]
```diff
@@ -167,18 +169,44 @@ export class LineEditor extends React.Component<LineEditorProps> {
...
+    const formItemLayout = {
+      labelCol: { span: 8 },
+      wrapperCol: { span: 16 }
+    };
+
+    const path = `LineEditor.${field}Field`;
+    const value = _get(symbolizer, `[${field}]`);
+    const component = CompositionUtil.handleComposition(composition, path, onChange, field, value, fallback);
+
+    if (component == null) {
```
Comment (KaiVolland) [inline src/Component/Symbolizer/LineEditor/LineEditor.tsx]: `===` could be more failsafe

File: src/Component/CompositionContext/CompositionContext.tsx [hunk 2]
```diff
@@ -0,0 +1,17 @@
...
+    dashPatternField?: React.ReactElement | false;
+    dashOffsetField?: React.ReactElement | false;
+    capField?: React.ReactElement | false;
+    JoinField?: React.ReactElement | false;
+    GraphicStrokeField?: React.ReactElement | false;
+    GraphicFillField?: React.ReactElement | false;
+  }
+}
+
+export const CompositionContext = React.createContext({});
```
Comment (KaiVolland) [inline src/Component/CompositionContext/CompositionContext.tsx]: missing newline

File: src/Util/CompositionUtil.ts [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+const _get = require('lodash/get');
+
+class CompositionUtil {
+
+  static handleComposition(composition: Compositions, path: string, onChange: Function, field: string, value: any, fallback: React.ReactElement): React.ReactElement {
+    const compositionValue = _get(composition, path);
+    const globalPath = path.split('.')[path.split('.').length - 1];
+    const globalComposition = _get(composition, globalPath);
+
+    if (compositionValue != undefined) {
```
Comment (KaiVolland) [inline src/Util/CompositionUtil.ts]:
```suggestion
    if (compositionValue !== undefined) {
```

File: src/Util/CompositionUtil.ts [hunk 4]
```diff
@@ -0,0 +1,37 @@
...
+      if (compositionValue != false) {
```
Comment (KaiVolland) [inline src/Util/CompositionUtil.ts]:
```suggestion
      if (compositionValue !== false) {
```

File: src/Component/Symbolizer/LineEditor/LineEditor.tsx [hunk 5]
```diff
@@ -167,18 +169,44 @@ export class LineEditor extends React.Component<LineEditorProps> {
     }
   }
 
+  handleComposition = (composition: Compositions, field:string, onChange: Function, fallback: React.ReactElement) => {
```
Comment (KaiVolland) [inline src/Component/Symbolizer/LineEditor/LineEditor.tsx]: I would be prefer `defaultElement` over `fallback`.

File: src/Component/Symbolizer/LineEditor/LineEditor.tsx [hunk 6]
```diff
@@ -194,93 +222,58 @@ export class LineEditor extends React.Component<LineEditorProps> {
...
-          </Panel>
-        </Collapse>
-      </div>
-    );
+          <CompositionContext.Consumer>
+            {(composition: Compositions) => (
+                <div className="gs-line-symbolizer-editor" >
+                  <Collapse bordered={false} defaultActiveKey={['1']} onChange={(key: string) => (null)}>
+                    <Panel header="General" key="1">
+                      {this.handleComposition(composition, 'color', this.onColorChange, <ColorField />)}
```
Comment (KaiVolland) [inline src/Component/Symbolizer/LineEditor/LineEditor.tsx]: I think it woul be nice to extract these lines to `const` above the `return` to keep the render tree as clean as possible.

**Comment by KaiVolland** [comment]
I think this is a good approach. The components get more complex but this is not avoidable.

If we document this well in the API-docs it should be very easy for users top make use of the customization.

Is this somewhere API breaking or would it be possible to merge this and proceed with further configurations on other components later?

**Comment by jansule** [comment]
> I think this is a good approach. The components get more complex but this is not avoidable.

I think so too. Complexity does not even increase too much, as we handle the main decision making in the utility class.

> If we document this well in the API-docs it should be very easy for users to make use of the customization.

This definitely needs to be done, as we are working with an anti-pattern here. And also users should know about this nice feature.

 
> Is this somewhere API breaking or would it be possible to merge this and proceed with further configurations on other components later?

It is NOT breaking the API, since we provide a default context. Everything stays the same if no context provider was set.

File: src/Component/Symbolizer/LineEditor/LineEditor.tsx (hunk 6 above)
Comment (jansule) [reply src/Component/Symbolizer/LineEditor/LineEditor.tsx]: Currently I do not think this is possible. The `CompositionContext.Consumer` component injects the context to its children. Right now, I do not see another way of accessing the context inside the render function. The old way of retrieving context (`this.context`) does not return the context, and the new way (`useContext(<Context>)` hook) does not work for classes.

File: src/Component/CompositionContext/CompositionContext.tsx (hunk 2 above)
Comment (jansule) [reply src/Component/CompositionContext/CompositionContext.tsx]: solved in https://github.com/terrestris/geostyler/pull/1186/commits/4fcbbd578e14eb8c7466107a521dd20ed8a860f5

File: src/Component/Symbolizer/LineEditor/LineEditor.tsx (hunk 5 above)
Comment (jansule) [reply src/Component/Symbolizer/LineEditor/LineEditor.tsx]: solved in https://github.com/terrestris/geostyler/pull/1186/commits/4fcbbd578e14eb8c7466107a521dd20ed8a860f5

File: src/Component/Symbolizer/Editor/Editor.tsx [hunk 7]
```diff
@@ -95,7 +97,7 @@ export class Editor extends React.Component<EditorProps, EditorState> {
     }
   }
 
-  getUiFromSymbolizer = (symbolizer: Symbolizer): React.ReactNode => {
+  getUiFromSymbolizer = (symbolizer: Symbolizer, composition: Compositions): React.ReactNode => {
```
Comment (KaiVolland) [inline src/Component/Symbolizer/Editor/Editor.tsx]: Please add some Code docs to this and the other newly introduced functions.

File: src/Util/CompositionUtil.spec.tsx [hunk 8]
```diff
@@ -0,0 +1,116 @@
...
+        propValue: dummyValue,
+        defaultElement: dummyDefaultElement
+      });
+
+      expect(composition).toBeDefined();
+      expect(composition).toBeFalsy();
+
+    });
+  });
+});
```
Comment (KaiVolland) [inline src/Util/CompositionUtil.spec.tsx]: newline please

File: src/Util/CompositionUtil.ts [hunk 9]
```diff
@@ -0,0 +1,86 @@
...
+
+interface CompositionUtilInjectOptions {
+  component: React.ReactElement;
+  onChange: Function;
+  propName: string;
+  propValue: any;
+  onChangeName?: string;
+}
+
+class CompositionUtil {
```
Comment (KaiVolland) [inline src/Util/CompositionUtil.ts]: Please add code docs.

File: src/Component/CompositionContext/CompositionContext.tsx [hunk 10]
```diff
@@ -0,0 +1,71 @@
+import * as React from 'react';
```
Comment (KaiVolland) [inline src/Component/CompositionContext/CompositionContext.tsx]: Please add some code docs what this file is about.

File: src/Component/Symbolizer/Editor/Editor.tsx (hunk 7 above)
Comment (jansule) [reply src/Component/Symbolizer/Editor/Editor.tsx]: solved in https://github.com/geostyler/geostyler/pull/1186/commits/587b9b93e34174a44cce9a8464bcb50512687670

File: src/Util/CompositionUtil.spec.tsx (hunk 8 above)
Comment (jansule) [reply src/Util/CompositionUtil.spec.tsx]: solved in https://github.com/geostyler/geostyler/pull/1186/commits/587b9b93e34174a44cce9a8464bcb50512687670

File: src/Util/CompositionUtil.ts (hunk 9 above)
Comment (jansule) [reply src/Util/CompositionUtil.ts]: solved in https://github.com/geostyler/geostyler/pull/1186/commits/587b9b93e34174a44cce9a8464bcb50512687670

File: src/Component/CompositionContext/CompositionContext.tsx (hunk 10 above)
Comment (jansule) [reply src/Component/CompositionContext/CompositionContext.tsx]: solved in https://github.com/geostyler/geostyler/pull/1186/commits/587b9b93e34174a44cce9a8464bcb50512687670

**Comment by jansule** [comment]
@KaiVolland added docs, please review again.

**Comment by KaiVolland** [comment]
Nice job! Please merge at will.

**Comment by jansule** [comment]
closes #650

**Pull request merged**